snap/info_snap_yaml.go

```diff
@@ -21,18 +21,198 @@ package snap
 
 import (
 	"fmt"
+	"strings"
+
 	"gopkg.in/yaml.v2"
 )
 
+type snapYaml struct {
+	Name        string                 `yaml:"name"`
+	Developer   string                 `yaml:"developer"`
+	Version     string                 `yaml:"version"`
+	Type        Type                   `yaml:"type"`
+	Channel     string                 `yaml:"channel"`
+	Description string                 `yaml:"description"`
+	Plugs       map[string]interface{} `yaml:"plugs,omitempty"`
+	Slots       map[string]interface{} `yaml:"slots,omitempty"`
+	Apps        map[string]appYaml     `yaml:"apps,omitempty"`
+}
+
+type plugYaml struct {
+	Interface string                 `yaml:"interface"`
+	Attrs     map[string]interface{} `yaml:"attrs,omitempty"`
+	Apps      []string               `yaml:"apps,omitempty"`
+	Label     string                 `yaml:"label"`
+}
+
+type slotYaml struct {
+	Interface string                 `yaml:"interface"`
+	Attrs     map[string]interface{} `yaml:"attrs,omitempty"`
+	Apps      []string               `yaml:"apps,omitempty"`
+	Label     string                 `yaml:"label"`
+}
+
+type appYaml struct {
+	Command   string   `yaml:"command"`
+	SlotNames []string `yaml:"slots,omitempty"`
+	PlugNames []string `yaml:"plugs,omitempty"`
+}
+
 // InfoFromSnapYaml creates a new info based on the given snap.yaml data
 func InfoFromSnapYaml(yamlData []byte) (*Info, error) {
-	var s Info
-	err := yaml.Unmarshal(yamlData, &s)
+	var y snapYaml
+	err := yaml.Unmarshal(yamlData, &y)
 	if err != nil {
 		return nil, fmt.Errorf("info failed to parse: %s", err)
 	}
-
+	// Construct snap skeleton, without apps, plugs and slots
+	snap := &Info{
+		Name:        y.Name,
+		Developer:   y.Developer,
+		Version:     y.Version,
+		Type:        y.Type,
+		Channel:     y.Channel,
+		Description: y.Description,
+		Apps:        make(map[string]*AppInfo),
+		Plugs:       make(map[string]*PlugInfo),
+		Slots:       make(map[string]*SlotInfo),
+	}
+	// Collect top-level definitions of plugs
+	for name, data := range y.Plugs {
+		iface, attrs, err := convertToSlotOrPlugData("plug", name, data)
+		if err != nil {
+			return nil, err
+		}
+		snap.Plugs[name] = &PlugInfo{
+			Snap:      snap,
+			Name:      name,
+			Interface: iface,
+			Attrs:     attrs,
+			Apps:      make(map[string]*AppInfo),
+		}
+	}
+	// Collect top-level definitions of slots
+	for name, data := range y.Slots {
+		iface, attrs, err := convertToSlotOrPlugData("slot", name, data)
+		if err != nil {
+			return nil, err
+		}
+		snap.Slots[name] = &SlotInfo{
+			Snap:      snap,
+			Name:      name,
+			Interface: iface,
+			Attrs:     attrs,
+			Apps:      make(map[string]*AppInfo),
+		}
+	}
+	// Collect app-level implicit definitions of plugs and slots
+	for _, yApp := range y.Apps {
```

Comment:
There are three loops on y.Apps, and it looks like we need a single one. Inside this one:

* Create the app
* Iterate over its plug names
* If non-existing, create plug
* Add app to plug, plug to app
* Same for slots